### PR TITLE
Improve Rate Limiter and Add tests.

### PR DESCRIPTION
**Pull request by zeeshanakram3**
- [x]  Updated Trust Pallet and Rate Limiting pallet to Substrate v3.0.0
- [x] Implemented sliding window rate limiting algorithm for free calls.
- [x] Adds tests for Rate Limiter and Trust pallets. 

#131

### CODE REVIEW

File: pallets/trust/src/mock.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		System: system::{Module, Call, Config, Storage, Event<T>},
+		Balances: pallet_balances::{Module, Call, Storage, Config<T>, Event<T>},
+		Utils: pallet_utils::{Module, Event<T>},
+		Trust: trust::{Module, Call, Storage, Event<T>},
+	}
+);
+
+parameter_types! {
+	pub const BlockHashCount: u64 = 250;
+	pub const SS58Prefix: u8 = 28;
```
Comment (F3Joule) [inline pallets/trust/src/mock.rs:31]:
```suggestion
```

File: runtime/Cargo.toml [hunk 2]
```diff
@@ -7,7 +7,13 @@ license = 'GPL-3.0-only'
...
 repository = 'https://github.com/dappforce/dappforce-subsocial-node'
 description = 'Subsocial runtime'
-keywords = ['blockchain', 'cryptocurrency', 'social-network', 'news-feed', 'marketplace']
+keywords = [
+    'blockchain',
+    'cryptocurrency',
+    'social-network',
+    'news-feed',
+    'marketplace',
+]
```
Comment (F3Joule) [inline runtime/Cargo.toml:16]: extra formatting

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -2,26 +2,21 @@
...
 
 // Make the WASM binary available.
 #[cfg(feature = "std")]
 include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
 
-use sp_std::{
-	prelude::*,
-	collections::btree_map::BTreeMap,
-};
+use sp_std::{prelude::*, collections::btree_map::BTreeMap};
```
Comment (F3Joule) [inline runtime/src/lib.rs:11]: extra formatting

File: runtime/src/lib.rs [hunk 4]
```diff
@@ -2,26 +2,21 @@
...
 use sp_runtime::{
-    ApplyExtrinsicResult, generic, create_runtime_str, impl_opaque_keys,
-    transaction_validity::{TransactionValidity, TransactionSource},
-};
-use sp_runtime::traits::{
-    BlakeTwo256, Block as BlockT, NumberFor, AccountIdLookup
+	ApplyExtrinsicResult, generic, create_runtime_str, impl_opaque_keys,
+	transaction_validity::{TransactionValidity, TransactionSource},
 };
+use sp_runtime::traits::{BlakeTwo256, Block as BlockT, NumberFor, AccountIdLookup};
```
Comment (F3Joule) [inline runtime/src/lib.rs:19]: extra formatting

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -37,51 +32,48 @@ pub use pallet_timestamp::Call as TimestampCall;
...
+	weights::{
+		Weight, IdentityFee, DispatchClass,
+		constants::{BlockExecutionWeight, ExtrinsicBaseWeight, RocksDbWeight, WEIGHT_PER_SECOND},
+	},
 };
 use frame_system::{
-    EnsureRoot,
-    limits::{BlockWeights, BlockLength}
+	EnsureRoot,
+	limits::{BlockWeights, BlockLength},
```
Comment (F3Joule) [inline runtime/src/lib.rs:44]: extra formatting

File: runtime/src/lib.rs [hunk 6]
```diff
@@ -37,51 +32,48 @@ pub use pallet_timestamp::Call as TimestampCall;
...
 use static_assertions::const_assert;
 
 use pallet_permissions::SpacePermission;
 use pallet_posts::rpc::{FlatPost, FlatPostKind, RepliesByPostId};
 use pallet_profiles::rpc::FlatSocialAccount;
-use pallet_reactions::{
-	ReactionId,
-	ReactionKind,
-	rpc::FlatReaction,
-};
```
Comment (F3Joule) [inline runtime/src/lib.rs:61]: extra formatting

File: runtime/src/lib.rs [hunk 7]
```diff
@@ -370,8 +399,8 @@ impl pallet_roles::Config for Runtime {
 	type MaxUsersToProcessPerDeleteRole = MaxUsersToProcessPerDeleteRole;
 	type Spaces = Spaces;
 	type SpaceFollows = SpaceFollows;
-	type IsAccountBlocked = ()/*Moderation*/;
-	type IsContentBlocked = ()/*Moderation*/;
+	type IsAccountBlocked = ();
+	type IsContentBlocked = ();
```
Comment (F3Joule) [inline runtime/src/lib.rs:403]: leave the comments here

File: runtime/src/lib.rs [hunk 8]
```diff
@@ -328,7 +357,7 @@ impl pallet_posts::Config for Runtime {
 	type MaxCommentDepth = MaxCommentDepth;
 	type PostScores = Scores;
 	type AfterPostUpdated = PostHistory;
-	type IsPostBlocked = ()/*Moderation*/;
+	type IsPostBlocked = ();
```
Comment (F3Joule) [inline runtime/src/lib.rs:360]: leave a comment here

File: runtime/src/lib.rs [hunk 9]
```diff
@@ -429,21 +458,21 @@ impl pallet_spaces::Config for Runtime {
 	type SpaceFollows = SpaceFollows;
 	type BeforeSpaceCreated = SpaceFollows;
 	type AfterSpaceUpdated = SpaceHistory;
-	type IsAccountBlocked = ()/*Moderation*/;
-	type IsContentBlocked = ()/*Moderation*/;
+	type IsAccountBlocked = ();
+	type IsContentBlocked = ();
```
Comment (F3Joule) [inline runtime/src/lib.rs:462]: leave the comments here

File: runtime/src/lib.rs [hunk 10]
```diff
@@ -452,14 +481,15 @@ impl pallet_space_history::Config for Runtime {}
...
-        let is_force_transfer = matches!(c, Call::Balances(pallet_balances::Call::force_transfer(..)));
-        match *c {
-            Call::Balances(..) => is_set_balance || is_force_transfer,
-            _ => true,
-        }
-    }
+	fn filter(c: &Call) -> bool {
+		let is_set_balance = matches!(c, Call::Balances(pallet_balances::Call::set_balance(..)));
+		let is_force_transfer =
+			matches!(c, Call::Balances(pallet_balances::Call::force_transfer(..)));
```
Comment (F3Joule) [inline runtime/src/lib.rs:487]: extra formatting

**Review by F3Joule — COMMENTED** [review]
Needs extra check of a rate-limiter design.

File: pallets/df-integration-tests/Cargo.toml [hunk 11]
```diff
@@ -7,7 +7,13 @@ license = 'GPL-3.0-only'
...
 repository = 'https://github.com/dappforce/dappforce-subsocial-node'
 description = 'Integration tests'
-keywords = ['blockchain', 'cryptocurrency', 'social-network', 'news-feed', 'marketplace']
+keywords = [
+    'blockchain',
+    'cryptocurrency',
+    'social-network',
+    'news-feed',
+    'marketplace',
+]
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/Cargo.toml:16]: Done

[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/Cargo.toml]: Done

File: pallets/df-integration-tests/Cargo.toml [hunk 12]
```diff
@@ -36,10 +42,14 @@ std = [
...
     'pallet-utils/std',
+    'pallet-trust/std',
+    'pallet-rate-limiter/std',
 ]
 
 [dependencies]
-codec = { package = 'parity-scale-codec', version = '2.0.0', default-features = false, features = ['derive'] }
+codec = { package = 'parity-scale-codec', version = '2.0.0', default-features = false, features = [
+    'derive',
+] }
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/Cargo.toml]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 13]
```diff
@@ -10,11 +10,12 @@ mod tests {
...
-        assert_ok, assert_noop,
-        parameter_types,
-        dispatch::{DispatchResult, DispatchError},
-        storage::StorageMap,
-    };
+		assert_noop, assert_ok,
+		dispatch::{DispatchError, DispatchResult, DispatchResultWithPostInfo, Dispatchable},
+		parameter_types,
+		storage::StorageMap,
+		traits::Filter,
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 14]
```diff
@@ -27,10 +28,12 @@ mod tests {
     use pallet_profile_follows::Error as ProfileFollowsError;
     use pallet_reactions::{ReactionId, ReactionKind, PostReactionScores, Error as ReactionsError};
     use pallet_scores::ScoringAction;
-    use pallet_spaces::{SpaceById, SpaceUpdate, Error as SpacesError, SpacesSettings};
+    use pallet_spaces::{Error as SpacesError, Space, SpaceById, SpaceUpdate, SpacesSettings};
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 15]
```diff
@@ -270,9 +275,77 @@ mod tests {
         type DefaultAutoblockThreshold = DefaultAutoblockThreshold;
     }
 
+    parameter_types! {}
+
+    impl pallet_trust::Config for TestRuntime {
+        type Event = Event;
+        type SetTrustLevel = system::EnsureRoot<AccountId>;
+        type WeightInfo = SubstrateWeight<TestRuntime>;
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 16]
```diff
@@ -270,9 +275,77 @@ mod tests {
...
+        type Call = Call;
+        type CallFilter = ();
+        type RateConfigs = RateConfigs;
+        type TrustHandler = Trust;
+    }
+
     type AccountId = u64;
     type BlockNumber = u64;
 
+    pub(crate) const MINUTES: BlockNumber = 60 as BlockNumber;
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs:334]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 17]
```diff
@@ -411,6 +484,30 @@ mod tests {
             ext.execute_with(|| Self::add_space_with_custom_permissions(permissions));
             ext
         }
+
+        /// Custom ext configuration with SpaceId, Initial Block Number for free
+        /// calls and email verified for given account
+        pub fn build_with_space_and_set_email_verified_for(who: AccountId) -> TestExternalities {
+            let mut ext = Self::build();
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 18]
```diff
@@ -411,6 +484,30 @@ mod tests {
...
+            ext.execute_with( || {
+                Self::add_default_space();
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 19]
```diff
@@ -411,6 +484,30 @@ mod tests {
...
+                System::set_block_number(INITIAL_BLOCK_NUMBER);
+                assert_ok!(_set_email_verified(None, Some(who)));
+            });
+            ext
+        }  
+        
+        /// Custom ext configuration with SpaceId, Initial Block Number for free
+        /// calls and phone number verified for given account
+        pub fn build_with_space_and_set_phone_number_verified_for(who: AccountId) -> TestExternalities {
+            let mut ext = Self::build();
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 20]
```diff
@@ -411,6 +484,30 @@ mod tests {
...
+            ext.execute_with( || {
+                Self::add_default_space();
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Done

File: pallets/trust/src/mock.rs (hunk 1 above)
Comment (zeeshanakram3) [reply pallets/trust/src/mock.rs]: Done

File: runtime/Cargo.toml (hunk 2 above)
Comment (zeeshanakram3) [reply runtime/Cargo.toml]: Done

File: runtime/src/lib.rs (hunk 3 above)
Comment (zeeshanakram3) [reply runtime/src/lib.rs]: Done

File: runtime/src/lib.rs (hunk 10 above)
Comment (zeeshanakram3) [reply runtime/src/lib.rs]: Done

File: runtime/src/lib.rs (hunk 9 above)
Comment (zeeshanakram3) [reply runtime/src/lib.rs]: Done

File: runtime/src/lib.rs (hunk 8 above)
Comment (zeeshanakram3) [reply runtime/src/lib.rs]: Done

File: runtime/src/lib.rs (hunk 7 above)
Comment (zeeshanakram3) [reply runtime/src/lib.rs]: Done

File: pallets/df-integration-tests/src/lib.rs [hunk 21]
```diff
@@ -270,9 +275,77 @@ mod tests {
...
+        //         max_permits: 10
+        //   },
+        //   RateConfig {
+        //       period: 1 * HOURS,
+        //       max_permits: 20
+        //   },
+        //   RateConfig {
+        //       period: 1 * DAYS,
+        //       max_permits: 40
+        //   }
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply pallets/df-integration-tests/src/lib.rs]: Will be used for testing free calls with multiple window sizes.

File: pallets/rate-limiter/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,350 @@
...
+        #[weight = {
+            let dispatch_info = call.get_dispatch_info();
+            (
+                // TODO: use benchmarking for setting a weight
+                dispatch_info.weight.saturating_add(T::DbWeight::get().reads_writes(2, 1)),
+                DispatchClass::Normal,
+                Pays::No,
+            )
+        }]
+        // TODO should this be public or private
```
Comment (F3Joule) [inline pallets/rate-limiter/src/lib.rs:158]: This should be public.

File: pallets/trust/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+pub mod weights;
+pub use weights::*;
```
Comment (F3Joule) [inline pallets/trust/src/lib.rs:20]:
```suggestion
pub use weights::WeightInfo;
```

**Review by F3Joule — CHANGES_REQUESTED** [review]

File: pallets/rate-limiter/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,350 @@
...
+            } else if stats.last_window < current_window {
+                stats = reset_stats(0);
+            }
+
+            // Calculate the number of permits consumed using sliding window rate limiting
+            // algorithm.
+            let permits_used = ((window.period - current_window_elapsed) *
+                stats.previous_window_consumed_permits.into() /
+                window.period) +
+                stats.current_window_consumed_permits.into();
```
Comment (F3Joule) [inline pallets/rate-limiter/src/lib.rs:238]: Consider splitting into more primitive calculations. Furthermore, `Saturating` is good to use here.

**Pull request closed without merging**